### ISSUE TITLE
compute-gmm-gop: command not found

### ISSUE THREAD

**Issue by serkankavak** [jimbozhang/kaldi-gop#29]
I built with `./build.sh` and tried to run the example in egs/gop-compute but I had this error: 

```
$ ./run.sh 
local/compute-gmm-gop.sh --nj 1 --cmd run.pl data/eval data/lang exp/tri1 exp/eval_gop
local/compute-gmm-gop.sh: feature type is delta
local/compute-gmm-gop.sh: computing GOP in data/eval using model from exp/tri1, putting results in exp/eval_gop
run.pl: job failed, log is in exp/eval_gop/log/gop.1.log
```

The log file:

```
# compute-gmm-gop exp/eval_gop/tree exp/eval_gop/final.mdl data/lang/L.fst "ark,s,cs:apply-cmvn  --utt2spk=ark:data/eval/split1/1/utt2spk scp:data/eval/split1/1/cmvn.scp scp:data/eval/split1/1/feats.scp ark:- | add-deltas  ark:- ark:- |" "ark:utils/sym2int.pl --map-oov 221 -f 2- data/lang/words.txt data/eval/split1/1/text|" ark,t:exp/eval_gop/gop.1 
# Started at Wed Aug  5 17:15:35 CST 2020
#
bash: line 1: compute-gmm-gop: command not found
# Accounting: time=0 threads=1
# Ended (code 127) at Wed Aug  5 17:15:35 CST 2020, elapsed time 0 seconds
```

After, instead of using this part of example run.sh :
```
case $feat_type in
  delta) feats="ark,s,cs:apply-cmvn $cmvn_opts --utt2spk=ark:$sdata/JOB/utt2spk scp:$sdata/JOB/cmvn.scp scp:$sdata/JOB/feats.scp ark:- | add-deltas $delta_opts ark:- ark:- |";;
  lda) feats="ark,s,cs:apply-cmvn $cmvn_opts --utt2spk=ark:$sdata/JOB/utt2spk scp:$sdata/JOB/cmvn.scp scp:$sdata/JOB/feats.scp ark:- | splice-feats $splice_opts ark:- ark:- | transform-feats $srcdir/final.mat ark:- ark:- |"
    cp $srcdir/final.mat $srcdir/full.mat $dir
   ;;
  *) echo "$0: invalid feature type $feat_type" && exit 1;
esac

echo "$0: computing GOP in $data using model from $srcdir, putting results in $dir"

mdl=$srcdir/final.mdl
tra="ark:utils/sym2int.pl --map-oov $oov -f 2- $lang/words.txt $sdata/JOB/text|";
$cmd JOB=1:$nj $dir/log/gop.JOB.log compute-gmm-gop $dir/tree $dir/final.mdl $lang/L.fst "$feats" "$tra" "ark,t:$dir/gop.JOB" || exit 1;
```

I used debug code which is
```
apply-cmvn --utt2spk=ark:data/eval/split1/1/utt2spk scp:data/eval/split1/1/cmvn.scp scp:data/eval/split1/1/feats.scp ark:- | add-deltas ark:- ark,t:data/eval/feats.1.ark.txt
utils/sym2int.pl --map-oov `cat data/lang/oov.int` -f 2- data/lang/words.txt data/eval/text > data/eval/trans
compute-gmm-gop exp/tri1/tree exp/tri1/final.mdl data/lang/L.fst ark,t:data/eval/feats.1.ark.txt ark,t:data/eval/trans ark,t:gop.1
```

The output is:
```
$ ./run.sh 
local/compute-gmm-gop.sh --nj 1 --cmd run.pl data/eval data/lang exp/tri1 exp/eval_gop
local/compute-gmm-gop.sh: feature type is delta
local/compute-gmm-gop.sh: computing GOP in data/eval using model from exp/tri1, putting results in exp/eval_gop
local/compute-gmm-gop.sh: line 74: apply-cmvn: command not found
local/compute-gmm-gop.sh: line 74: add-deltas: command not found
sym2int.pl: replacing ORALS with 221
sym2int.pl: replacing P with 221
sym2int.pl: replacing H with 221
sym2int.pl: replacing D with 221
sym2int.pl: replacing (TW-) with 221
sym2int.pl: replacing YAHOO with 221
sym2int.pl: replacing (N-) with 221
sym2int.pl: replacing (AL-) with 221
sym2int.pl: replacing (YOU) with 221
sym2int.pl: replacing (MEAN) with 221
sym2int.pl: replacing MHM with 221
sym2int.pl: replacing (PLANES) with 221
sym2int.pl: replacing SCHOLARLY with 221
sym2int.pl: replacing (US-) with 221
sym2int.pl: replacing RACHEL with 221
sym2int.pl: replacing (C-) with 221
sym2int.pl: replacing (%HESITATION) with 221
sym2int.pl: replacing (WEE-) with 221
sym2int.pl: replacing CESSNA with 221
** Replaced 19 instances of OOVs with 221
local/compute-gmm-gop.sh: line 76: compute-gmm-gop: command not found
local/compute-gmm-gop.sh: done computing GOP.
```

**Comment by jimbozhang** [comment]
After building by `build.sh`, was the executable binary `src/gopbin/compute-gmm-gop` generated?

**Comment by serkankavak** [comment]
> After building by `build.sh`, was the executable binary `src/gopbin/compute-gmm-gop` generated?

No, but I think I solved the problem. The problem was building Kaldi 5.3 with CUDA 10.0. Kaldi 5.3 does not support CUDA 10.0 but the script builds Kaldi as CUDA enabled. I disabled it and now its working.

Just as in the previous issue #5 , when i run the example now i have 
`WARNING (compute-gmm-gop[5.1.113~1-72672]:Decode():gop/gmm-gop.cc:65) Did not successfully decode.`
It gives the output but I dont know how accurate it is. I am going to use my own model and give feedback here.

**Comment by amandeepbaberwal** [comment]
@serkankavak hi how did you build it by disabling Cuda? isn't the configuration file generate after building kaldi?

**Comment by serkankavak** [comment]
> @serkankavak hi how did you build it by disabling Cuda? isn't the configuration file generate after building kaldi?

[From Line 13 to 35](https://github.com/jimbozhang/kaldi-gop/blob/4e842a03946ac91e2363db5912ac973e2434c97c/src/CMakeLists.txt#L13) 
You can check the Kaldi installation configurations from [here](https://github.com/kaldi-asr/kaldi/blob/5.3/src/configure). `--use-cuda` is yes as default. I changed it to no by adding it to the makefile in [here](https://github.com/jimbozhang/kaldi-gop/blob/4e842a03946ac91e2363db5912ac973e2434c97c/src/CMakeLists.txt#L32)

Basically, replace `execute_process(COMMAND ./configure WORKING_DIRECTORY ${KALDI_ROOT}/src)` with `execute_process(COMMAND ./configure WORKING_DIRECTORY ${KALDI_ROOT}/src) --use-cuda=no`

**Comment by amandeepbaberwal** [comment]
@serkankavak i did this
execute_process(COMMAND ./configure --use-cuda=no WORKING_DIRECTORY ${KALDI_ROOT}/src)
--use-cuda=no supposed to be inside, but the problem still persists, when i checked configuration file it still says --use_cuda=true

log file says it can't find apply-cmvn and add-deltas, don't know what i am missing????
### log file, i don't know why some of it showing big on github
# compute-gmm-gop exp/eval_gop/tree exp/eval_gop/final.mdl data/lang/L.fst "ark,s,cs:apply-cmvn  --utt2spk=ark:data/eval/split1/1/utt2spk scp:data/eval/split1/1/cmvn.scp scp:data/eval/split1/1/feats.scp ark:- | add-deltas  ark:- ark:- |" "ark:utils/sym2int.pl --map-oov 221 -f 2- data/lang/words.txt data/eval/split1/1/text|" ark,t:exp/eval_gop/gop.1 
# Started at Thu Feb 23 22:31:43 IST 2023
#
compute-gmm-gop exp/eval_gop/tree exp/eval_gop/final.mdl data/lang/L.fst 'ark,s,cs:apply-cmvn  --utt2spk=ark:data/eval/split1/1/utt2spk scp:data/eval/split1/1/cmvn.scp scp:data/eval/split1/1/feats.scp ark:- | add-deltas  ark:- ark:- |' 'ark:utils/sym2int.pl --map-oov 221 -f 2- data/lang/words.txt data/eval/split1/1/text|' ark,t:exp/eval_gop/gop.1 
sh: 1: apply-cmvn: not found
sh: 1: add-deltas: not found
sym2int.pl: replacing ORALS with 221
sym2int.pl: replacing P with 221
sym2int.pl: replacing H with 221
sym2int.pl: replacing D with 221
sym2int.pl: replacing (TW-) with 221
sym2int.pl: replacing YAHOO with 221
sym2int.pl: replacing (N-) with 221
sym2int.pl: replacing (AL-) with 221
sym2int.pl: replacing (YOU) with 221
sym2int.pl: replacing (MEAN) with 221
sym2int.pl: replacing MHM with 221
sym2int.pl: replacing (PLANES) with 221
sym2int.pl: replacing SCHOLARLY with 221
sym2int.pl: replacing (US-) with 221
sym2int.pl: replacing RACHEL with 221
sym2int.pl: replacing (C-) with 221
sym2int.pl: replacing (%HESITATION) with 221
sym2int.pl: replacing (WEE-) with 221
sym2int.pl: replacing CESSNA with 221
** Replaced 19 instances of OOVs with 221
LOG (compute-gmm-gop[5.3]:main():gopbin/compute-gmm-gop.cc:68) Done.
WARNING (compute-gmm-gop[5.3]:Close():kaldi-io.cc:512) Pipe apply-cmvn  --utt2spk=ark:data/eval/split1/1/utt2spk scp:data/eval/split1/1/cmvn.scp scp:data/eval/split1/1/feats.scp ark:- | add-deltas  ark:- ark:- | had nonzero return status 32512
ERROR (compute-gmm-gop[5.3]:~SequentialTableReaderArchiveImpl():util/kaldi-table-inl.h:678) TableReader: error detected closing archive 'apply-cmvn  --utt2spk=ark:data/eval/split1/1/utt2spk scp:data/eval/split1/1/cmvn.scp scp:data/eval/split1/1/feats.scp ark:- | add-deltas  ark:- ark:- |'

[ Stack-Trace: ]

kaldi::MessageLogger::HandleMessage(kaldi::LogMessageEnvelope const&, char const*)
kaldi::MessageLogger::~MessageLogger()
kaldi::SequentialTableReaderArchiveImpl<kaldi::KaldiObjectHolder<kaldi::Matrix<float> > >::~SequentialTableReaderArchiveImpl()
kaldi::SequentialTableReaderArchiveImpl<kaldi::KaldiObjectHolder<kaldi::Matrix<float> > >::~SequentialTableReaderArchiveImpl()
kaldi::SequentialTableReader<kaldi::KaldiObjectHolder<kaldi::Matrix<float> > >::~SequentialTableReader()
main

__libc_start_main
_start

terminate called after throwing an instance of 'std::runtime_error'
  what():  
# Accounting: time=1 threads=1
# Ended (code 134) at Thu Feb 23 22:31:44 IST 2023, elapsed time 1 seconds

**Comment by serkankavak** [comment]
@amandeepbaberwal I don't know why you face with this error but i guess your kaldi installation was not correctly done. Some dependencies might be missing etc. I suggest you to delete all the files and follow the installation instructions. Also, read the logs in the terminal if there is any error or warning while installing.

**Comment by amandeepbaberwal** [comment]
@serkankavak it is building OK with no errors, can you confirm which os and version you used? It may be something with the latest ubuntu version, i am using ubuntu 22.05 
Thanks

**Comment by serkankavak** [comment]
@amandeepbaberwal I used Ubuntu 18.04

**Comment by amandeepbaberwal** [comment]
@serkankavak it worked when i build the kaldi 5.3 from scratch, i ran make from tools and src. 
I am complete noob to this, when i check gop under eval/ it shows very confusing text, do you know how can i convert this to a human readable text? 
I mean is this gop coming from after converting from an audio file? These is not much documentation in readme file

**Comment by serkankavak** [comment]
I don't remember the details. I used it for a specific project and I don't have access to the source code now. The reason why I used GMM-GOP was because of my dataset was not big enough to train a neural network. If you think your dataset is big then I strongly suggest you to follow [this project](https://github.com/kaldi-asr/kaldi/tree/master/egs/gop_speechocean762).

**Comment by amandeepbaberwal** [comment]
@serkankavak currently i am following this guide https://github.com/YuanGongND/gopt using speechocean data and librispeech

**Comment by amandeepbaberwal** [comment]
but i think the project you are saying is much better, currently i am trying to get my head around what kaldi is and probably get that project eventually.
